### ISSUE TITLE
Link error when using LLVM or Intel Toolset under Visual Studio

### ISSUE THREAD

**Issue by Baljak** [pybind/pybind11#927]
When using the last LLVM and Intel C++ Compiler versions (4.0 and 17.0 respectively) under Visual Studio, the following error occurs (both with 2.1.1 and the current master):

``
error LNK2001: unresolved external symbol "public: __cdecl pybind11::error_already_set::error_already_set(class pybind11::error_already_set const &)" (??0error_already_set@pybind11@@QEAA@AEBV01@@Z)
1>C:\Dev\PythonLibs\Tests\Pybind11\build\bin\Release\pybind11test.cp36-win_amd64.pyd : fatal error LNK1120: 1 unresolved externals
``

The following code was used:

```c++
#include <pybind11/pybind11.h>

namespace py = pybind11;

struct S
{
	int value;

	int get() const { return value; }
	void set(int v) { value = v; }
};


PYBIND11_MODULE(pybind11test, m)
{
	py::class_<S>(m, "S")
		.def(py::init<>())
		.def("get", &S::get)
		.def("set", &S::set)
		.def_readwrite("spectrumType", &S::value);
}
```

I was able to fix this problem by changing line 704 in common.h (from the master branch):
```c++
error_already_set(const error_already_set &) = delete;
```
to
```c++
error_already_set(const error_already_set &) = default;
```

But I'm not sure this is the right way to fix the problem though.

**Comment by jagerman** [comment]
This seems like a bug in the respective compilers.

Does it help if you make the `delete`d copy constructor `private` rather than `public`?  (I suppose at the very least that might help figure out where the copy is actually happening).

**Comment by Baljak** [comment]
When making the copy constructor `private`, I get the same error on both compilers with no additional information. I switched to LLVM 4.01 which was recently released with no more success.

**Comment by Baljak** [comment]
Ok, I think this a not a compiler bug and that Intel and Clang are write here. See https://stackoverflow.com/questions/16432959/exception-and-copy-constructor-c.

Especially:
§5:
`**When the thrown object is a class object, the copy/move constructor and the destructor shall be accessible, even if the copy/move operation is elided (12.8).**`

So the copy constructor of error_already_set should be either defaulted or actually implemented, but not deleted.

**Comment by Baljak** [comment]
Ok, I think this a not a compiler bug and that Intel and Clang are write here. See https://stackoverflow.com/questions/16432959/exception-and-copy-constructor-c.

Especially:
**When the thrown object is a class object, the copy/move constructor and the destructor shall be accessible, even if the copy/move operation is elided (12.8).**

So the copy constructor of error_already_set should be either defaulted or actually implemented, but not deleted.

**Comment by Baljak** [comment]
Ok, I think this a not a compiler bug and that Intel and Clang are write here. See https://stackoverflow.com/questions/16432959/exception-and-copy-constructor-c.

Especially:
**When the thrown object is a class object, the copy/move constructor and the destructor shall be accessible, even if the copy/move operation is elided (12.8).**

So the copy constructor of error_already_set should be either defaulted or actually implemented, but not deleted.

**Comment by jagerman** [comment]
The thing about this that makes me thing it's a compiler bug is that with the explicit `= delete;` there should be a compilation failure if a copy constructor is required, not a linking failure.  The compiler is essentially acting as though the `= delete;` isn't there at all.  This is a bit weird: each of MSVC, clang, and ICC can handle this code just without any such error on their own; but for some reason ICC and clang are doing something that makes the MSVC linker want to link to the non-existent copy constructor.

I tried (and failed) to get the test suite building under MSVC/clang-cl.exe/cmake.  I'm not sure if it's an issue with VS 2017, LLVM, cmake, or me, but I couldn't get the test suite compiled (it seems the cmake scripts adds `-fexceptions -frtti` to the compiler flags, but `clang-cl.exe` barfs on those, looking instead for the MSVC `cl.exe` flags; at that point I gave up.)

Now all of that said, I took a better look at `error_already_set` and I think it's considerably more complicated than it needs to be.  I've submitted PR #954 to simplify it and, hopefully, fix this issue—given my compilation trouble above, feedback as to whether this works would be appreciated.
